lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -0,0 +1,691 @@
+    Operation *post = *loadOp.getResult().getUsers().begin();
+    auto newOrder = post->getResult(0)
+                        .getType()
+                        .cast<RankedTensorType>()
+                        .getEncoding()
+                        .cast<ttg::SharedEncodingAttr>()
+                        .getOrder();
+    auto ty = loadOp.getType().cast<RankedTensorType>();
+    auto oldOrder = ttg::getOrder(ty.getEncoding());
+    // The operand of MMAv3 is in SharedEncoding and it's order should not
+    // be changed after FuseTranspositions Pass. So we only pipeline the
+    // load if the order of the loaded BlockedEncoding is the same as the
+    // order of the SharedEncoding it is converted to.
+    // TODO: remove this constraint once the LoadOp supports transpose
+    // fusion
+    if (newOrder[0] == oldOrder[0] || newOrder[1] == oldOrder[1]) {
+      hasMMAV3 = true;
+      return preUse->getResult(0);
+    }
+  }
+  return Value();
+}
+
+namespace {
+struct LoadDotOperand {
+  LoadDotOperand(tt::LoadOp load, Value dotOperand)
+      : load(load), dotOperand(dotOperand) {}
+  tt::LoadOp load;
+  Value dotOperand;
+};
+} // namespace
+
+/// Collect loads to pipeline. Return success if we can pipeline this loop
+static void collectOpsToPipeline(scf::ForOp forOp,
+                                 SmallVectorImpl<LoadDotOperand> &ops,
+                                 bool &hasMMAV3) {
+  ModuleOp moduleOp = forOp->getParentOfType<ModuleOp>();
+  ModuleAxisInfoAnalysis axisInfoAnalysis(moduleOp);
+
+  // We cannot use forOp.walk(...) here because we only want to visit the
+  // operations in the loop body block. Nested blocks are handled separately.
+  for (Operation &op : forOp) {
+    if (auto loadOp = dyn_cast<tt::LoadOp>(&op)) {
+      bool candidate = false;
+      if (isLoadFromTensorPtr(loadOp)) {
+        // TODO: enable TMA pipelining.
+        candidate = false;
+      } else {
+        auto ptr = loadOp.getPtr();
+        unsigned vec = axisInfoAnalysis.getPtrContiguity(ptr);
+        if (auto mask = loadOp.getMask())
+          vec =
+              std::min<unsigned>(vec, axisInfoAnalysis.getMaskAlignment(mask));
+
+        auto tensorTy = ptr.getType().dyn_cast<RankedTensorType>();
+        if (!tensorTy || tensorTy.getRank() < 2)
+          continue;
+        auto ty =
+            tensorTy.getElementType().cast<tt::PointerType>().getPointeeType();
+        unsigned width = vec * ty.getIntOrFloatBitWidth();
+        // We do not pipeline all loads for the following reasons:
+        // 1. On nvidia GPUs, cp.async's cp-size can only be 4, 8 and 16.
+        // 2. It's likely that pipling small loads won't offer much performance
+        //    improvement and may even hurt performance by increasing register
+        //    pressure.
+        if (width >= 32)
+          candidate = true;
+      }
+      if (!candidate)
+        continue;
+      Value dotOperand = loadDotOperand(loadOp, hasMMAV3);
+      if (!dotOperand)
```

Comment:
Do we only pipeline dotOp loads ?

Reply:
I guess this is just to be on par with the current pipeliner, but do you think there will be an issue relying on LoadDotOp struct if we want to extend for other loads ?

Reply:
Right this is to match existing level of support. We can pipeline whatever we want as long as we pick a schedule for it and generate the async wait correct. 

>do you think there will be an issue relying on LoadDotOp struct if we want to extend for other loads ?
The goal of the refactor is to allow this kind of cases so I don't expect any issue but there may be some rough edges once we start doing it and we should improve things based on that.

Reply:
Nevermind, I think it should be fine